### PR TITLE
Allow to free direct buffers on java9 again

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

Java9 adds a new method to Unsafe which allows to free direct ByteBuffer via the cleaner without the need to use an commandline arguments.

Modifications:

- Add Cleaner interface
- Add CleanerJava9 which will be used when using Java9+ and take care of release direct ByteBuffer
- Let Cleaner0 implement Cleaner

Result:

Be able to free direct ByteBuffer on Java9+ again without any commandline arguments.

### CODE REVIEW

File: common/src/main/java/io/netty/util/internal/PlatformDependent0.java [hunk 1]
```diff
@@ -281,10 +282,6 @@ public Object run() {
 
         logger.debug("java.nio.DirectByteBuffer.<init>(long, int): {}",
                 DIRECT_BUFFER_CONSTRUCTOR != null ? "available" : "unavailable");
-
-        if (direct != null) {
-            freeDirectBuffer(direct);
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/PlatformDependent0.java]: removing this makes it easier to break cycle references and its not super important anyway

File: common/src/main/java/io/netty/util/internal/Cleaner0.java [hunk 2]
```diff
@@ -29,10 +29,9 @@
  *
  * For more details see <a href="https://github.com/netty/netty/issues/2604">#2604</a>.
  */
-final class Cleaner0 {
+final class Cleaner0 implements Cleaner {
```
Comment (netkins) [inline common/src/main/java/io/netty/util/internal/Cleaner0.java]: CRITICAL Initialization circularity between io.netty.util.internal.Cleaner0 and io.netty.util.internal.CleanerJava9 [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=findbugs%3AIC_INIT_CIRCULARITY)
CRITICAL Initialization circularity between io.netty.util.internal.Cleaner0 and io.netty.util.internal.PlatformDependent [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=findbugs%3AIC_INIT_CIRCULARITY)
CRITICAL Initialization circularity between io.netty.util.internal.Cleaner0 and io.netty.util.internal.PlatformDependent0 [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=findbugs%3AIC_INIT_CIRCULARITY)

File: common/src/main/java/io/netty/util/internal/CleanerJava9.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+import sun.misc.Unsafe;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.nio.ByteBuffer;
+
+/**
+ * Provide a way to clean a ByteBuffer on Java9+.
+ */
+final class CleanerJava9 implements Cleaner {
```
Comment (netkins) [inline common/src/main/java/io/netty/util/internal/CleanerJava9.java]: CRITICAL Initialization circularity between io.netty.util.internal.CleanerJava9 and io.netty.util.internal.PlatformDependent [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=findbugs%3AIC_INIT_CIRCULARITY)
CRITICAL Initialization circularity between io.netty.util.internal.CleanerJava9 and io.netty.util.internal.PlatformDependent0 [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=findbugs%3AIC_INIT_CIRCULARITY)

**Comment by johnou** [comment]
@normanmaurer how about the non unsafe version? https://bugs.openjdk.java.net/browse/JDK-8148117

**Comment by normanmaurer** [comment]
@johnou not sure i get your question... Can you explain ?

File: common/src/main/java/io/netty/util/internal/Cleaner0.java (hunk 2 above)
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/Cleaner0.java]: I think its fine... @Scottmitch @nmittler @trustin PTAL

File: common/src/main/java/io/netty/util/internal/CleanerJava9.java (hunk 3 above)
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/CleanerJava9.java]: I think its fine... @Scottmitch @nmittler @trustin PTAL

**Comment by johnou** [comment]
@normanmaurer nvm, I misread "If, at some point in the future, it is determined that sun.misc.Cleaner is in fact a critical internal API ( with static usage ), 
then it can be reinstated as a subtype of jdk.internal.ref.Cleaner, providing the same public API. " as though there would already be a public API.

**Comment by normanmaurer** [comment]
@trustin PTAL as well

File: common/src/main/java/io/netty/util/internal/Cleaner0.java (hunk 2 above)
Comment (fenik17) [inline common/src/main/java/io/netty/util/internal/Cleaner0.java]: It's a possible deadlock such as [guava issue #1977](https://github.com/google/guava/issues/1977)?

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/Cleaner0.java]: I think I found the cycle... stay tuned

File: common/src/main/java/io/netty/util/internal/CleanerJava9.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+* distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+* WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+* License for the specific language governing permissions and limitations
+* under the License.
+*/
+package io.netty.util.internal;
+
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+import sun.misc.Unsafe;
```
Comment (netkins) [inline common/src/main/java/io/netty/util/internal/CleanerJava9.java]: MINOR Remove this unused import 'sun.misc.Unsafe'. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: common/src/main/java/io/netty/util/internal/CleanerJava9.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+* distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+* WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+* License for the specific language governing permissions and limitations
+* under the License.
+*/
+package io.netty.util.internal;
+
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
```
Comment (netkins) [inline common/src/main/java/io/netty/util/internal/CleanerJava9.java]: MINOR Remove this unused import 'sun.misc.Unsafe'. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: common/src/main/java/io/netty/util/internal/Cleaner0.java (hunk 2 above)
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/Cleaner0.java]: fixed

File: common/src/main/java/io/netty/util/internal/PlatformDependent0.java [hunk 6]
```diff
@@ -244,7 +246,7 @@ public Object run() {
                 public Object run() {
                     try {
                         Class<?> bitsClass =
-                                Class.forName("java.nio.Bits", false, PlatformDependent.getSystemClassLoader());
+                                Class.forName("java.nio.Bits", false, getSystemClassLoader());
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/PlatformDependent0.java]: This fixed a cycle reference

File: common/src/main/java/io/netty/util/internal/PlatformDependent0.java [hunk 7]
```diff
@@ -281,10 +283,35 @@ public Object run() {
...
                 DIRECT_BUFFER_CONSTRUCTOR != null ? "available" : "unavailable");
+    }
+
+    static boolean isExplicitNoUnsafe() {
+        return IS_EXPLICIT_NO_UNSAFE;
+    }
 
-        if (direct != null) {
-            freeDirectBuffer(direct);
+    private static boolean explicitNoUnsafe0() {
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/PlatformDependent0.java]: I moved this from `PlatformDependent` to remove the a cycle reference.

File: common/src/main/java/io/netty/util/internal/CleanerJava9.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+            method = null;
+            logger.debug("java.nio.ByteBuffer.cleaner(): unavailable", (Throwable) maybeInvokeMethod);
+        } else {
+            method = (Method) maybeInvokeMethod;
+            logger.debug("java.nio.ByteBuffer.cleaner(): available");
+        }
+        INVOKE_CLEANER = method;
+    }
+
+    private static void freeDirectBuffer0(ByteBuffer buffer) {
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/CleanerJava9.java]: is this method necessary? seems like it is only used in a single place

File: common/src/main/java/io/netty/util/internal/Cleaner0.java [hunk 9]
```diff
@@ -72,31 +63,28 @@
...
+        freeDirectBuffer0(direct);
+    }
+
+    @Override
+    public void freeDirectBuffer(ByteBuffer buffer) {
+        freeDirectBuffer0(buffer);
     }
 
-    static void freeDirectBuffer(ByteBuffer buffer) {
+    private static void freeDirectBuffer0(ByteBuffer buffer) {
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/Cleaner0.java]: I think we can remove this method, and just move the code into `freeDirectBuffer`. We invoke the cleaner above `clean.invoke(cleaner);` so we may be double freeing this buffer

File: common/src/main/java/io/netty/util/internal/PlatformDependent.java [hunk 10]
```diff
@@ -122,7 +120,7 @@ public Random current() {
             logger.debug("-Dio.netty.noPreferDirect: {}", !DIRECT_BUFFER_PREFERRED);
         }
 
-        if (!hasUnsafe() && !isAndroid() && !IS_EXPLICIT_NO_UNSAFE) {
+        if (!hasUnsafe() && !isAndroid() && !PlatformDependent0.isExplicitNoUnsafe()) {
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/PlatformDependent.java]: `isExplicitNoUnsafe` should be reflected in `hasUnsafe` ... can we just remove the `!PlatformDependent0.isExplicitNoUnsafe()` part?

File: common/src/main/java/io/netty/util/internal/PlatformDependent.java [hunk 11]
```diff
@@ -158,6 +156,19 @@ public Random current() {
         logger.debug("io.netty.maxDirectMemory: {} bytes", maxDirectMemory);
 
         MAYBE_SUPER_USER = maybeSuperUser0();
+
+        if (hasUnsafe() && !isAndroid()) {
+            // only direct to method if we are not running on android.
+            // See https://github.com/netty/netty/issues/2604
+            CLEANER = javaVersion() >= 9 ? new CleanerJava9() : new Cleaner0();
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/PlatformDependent.java]: consider renaming `Cleaner0` to `CleanerJava6`.

**Comment by normanmaurer** [comment]
@Scottmitch addressed all of it... Good to merge ?

File: common/src/main/java/io/netty/util/internal/CleanerJava6.java [hunk 12]
```diff
@@ -29,39 +29,29 @@
...
+                clean = cleaner.getClass().getDeclaredMethod("clean");
+                clean.invoke(cleaner);
             } catch (Throwable t) {
                 // We don't have ByteBuffer.cleaner().
                 fieldOffset = -1;
                 clean = null;
-                cleanerIsRunnable = false;
                 error = t;
             }
         }
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/CleanerJava6.java]: the conditional below checks if `error == null` and then says "cleaner is available". However this may not be the case if we have no unsafe. We should adjust the logic below or set the `error = <exception indicating unsafe isn't available>`.

File: common/src/main/java/io/netty/util/internal/CleanerJava6.java [hunk 13]
```diff
@@ -72,31 +62,21 @@
...
 
-    static void freeDirectBuffer(ByteBuffer buffer) {
+    @Override
+    public void freeDirectBuffer(ByteBuffer buffer) {
         if (CLEANER_FIELD_OFFSET == -1 || !buffer.isDirect()) {
             return;
         }
-        assert CLEAN_METHOD != null || CLEANER_IS_RUNNABLE:
-                "CLEANER_FIELD_OFFSET != -1 implies CLEAN_METHOD != null or CLEANER_IS_RUNNABLE == true";
+        assert CLEAN_METHOD != null;
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/CleanerJava6.java]: nit: remove this assert ... we will very likely NPE via a JVM trap below anyways.

File: common/src/main/java/io/netty/util/internal/CleanerJava6.java [hunk 14]
```diff
@@ -72,31 +62,21 @@
...
             if (cleaner != null) {
-                if (CLEANER_IS_RUNNABLE) {
-                    ((Runnable) cleaner).run();
-                } else {
-                    CLEAN_METHOD.invoke(cleaner);
-                }
+                CLEAN_METHOD.invoke(cleaner);
             }
         } catch (Throwable t) {
             // Nothing we can do here.
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/CleanerJava6.java]: can we log a debug statement, or is this expected to happen frequently?

File: common/src/main/java/io/netty/util/internal/CleanerJava9.java [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+            maybeInvokeMethod = e;
+        }
+        if (maybeInvokeMethod instanceof Throwable) {
+            method = null;
+            logger.debug("java.nio.ByteBuffer.cleaner(): unavailable", (Throwable) maybeInvokeMethod);
+        } else {
+            method = (Method) maybeInvokeMethod;
+            logger.debug("java.nio.ByteBuffer.cleaner(): available");
+        }
+        INVOKE_CLEANER = method;
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/CleanerJava9.java]: we should also log something similar to the `Java6` class if the cleaner is not supported.

File: common/src/main/java/io/netty/util/internal/CleanerJava9.java [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    @Override
+    public void freeDirectBuffer(ByteBuffer buffer) {
+        if (INVOKE_CLEANER == null) {
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/CleanerJava9.java]: suggestion: instead of checking this condition every time we could have `PlatformDependent` detect if the cleaner is properly initialized, and if not just use the NOOP cleaner. We should do the same thing for the java6 class:

```java
class CleanerJava9 {

static {
..
}

static boolean isInitialized() {
  return INVOKE_CLEANER != null;
}
..
}

class PlatformDependent {
..
if (java9) {
  CLEANER = CleanerJava9.isInitialized() ? new CleanerJava9() : NOOP;
} else {
  CLEANER = CleanerJava6.isInitialized() ? new CleanerJava6() : NOOP;
}
...
}
```

File: common/src/main/java/io/netty/util/internal/CleanerJava6.java (hunk 14 above)
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/CleanerJava6.java]: i would argue this should never happen as we tried it before in the static block. Maybe best to throw an Error . WDYT ?

Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/CleanerJava6.java]: yah if we never expect it to happen lets re-throw it. What about `PlatformDependent.throwException(t);`?

File: common/src/main/java/io/netty/util/internal/CleanerJava6.java [hunk 17]
```diff
@@ -72,31 +64,24 @@
         }
         CLEANER_FIELD_OFFSET = fieldOffset;
         CLEAN_METHOD = clean;
-        CLEANER_IS_RUNNABLE = cleanerIsRunnable;
+    }
 
-        // free buffer if possible
-        freeDirectBuffer(direct);
+    public static boolean isSupported() {
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/CleanerJava6.java]: package private?

File: common/src/main/java/io/netty/util/internal/CleanerJava9.java [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+        }
+        if (error == null) {
+            logger.debug("java.nio.ByteBuffer.cleaner(): available");
+        } else {
+            logger.debug("java.nio.ByteBuffer.cleaner(): unavailable", error);
+        }
+        INVOKE_CLEANER = method;
+    }
+
+    public static boolean isSupported() {
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/CleanerJava9.java]: package private?

File: common/src/main/java/io/netty/util/internal/CleanerJava9.java [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+/**
+ * Provide a way to clean a ByteBuffer on Java9+.
+ */
+final class CleanerJava9 implements Cleaner {
+    private static final InternalLogger logger = InternalLoggerFactory.getInstance(CleanerJava9.class);
+
+    private static final Method INVOKE_CLEANER;
+
+    static {
+        Method method;
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/CleanerJava9.java]: final?

File: common/src/main/java/io/netty/util/internal/CleanerJava9.java [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+        Throwable error;
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/CleanerJava9.java]: final?

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (7b6119a0a478b40f12e85c0e92e603f9f6ee5b6c) and 4.0 (295c24e246470dfdcc63066133ebcaa23c5144a9)

File: common/src/main/java/io/netty/util/internal/PlatformDependent.java [hunk 21]
```diff
@@ -122,7 +127,7 @@ public Random current() {
             logger.debug("-Dio.netty.noPreferDirect: {}", !DIRECT_BUFFER_PREFERRED);
         }
 
-        if (!hasUnsafe() && !isAndroid() && !IS_EXPLICIT_NO_UNSAFE) {
```
Comment (jasontedor) [inline common/src/main/java/io/netty/util/internal/PlatformDependent.java]: Why was this removed?  It means that logging for this shows up even if no unsafe was explicitly set which was the point of adding this in the first place It means that with Netty 4.1.10.Final our users are back to seeing a message that they don't understand that makes them think their system is broken. Unless there is a good reason that this is a removed (I'm sorry if I'm missing it), I would like to open a PR to add it back.

Comment (jasontedor) [inline common/src/main/java/io/netty/util/internal/PlatformDependent.java]: Relates #5624

**Comment by normanmaurer** [comment]
There was no good reason and it's my fault to miss this :(

Can you please do a PR and also add a comment in the code why it should never be removed again.

Sorry again for the trouble

File: common/src/main/java/io/netty/util/internal/PlatformDependent.java (hunk 21 above)
Comment (jasontedor) [inline common/src/main/java/io/netty/util/internal/PlatformDependent.java]: Also, it seems to me that this change was not needed for the stated purpose of this change (to allow freeing direct buffers on JDK 9)? Again, maybe I'm missing something and if so I'm sorry.

Adding unrelated changes to PRs makes it hard to follow what is going on. I follow this repository, if I see a change titled "Allow to free direct buffers on java9 again" I think I know what it's about and I skip reading the code change. If an unrelated code change is slipped in, how can I as a follower know this without reading every code change which I think is an unreasonable assumption?

Had this been in a separate PR I would have seen it and would have responded then to please not remove it.

**Comment by jasontedor** [comment]
Thank you for understanding @normanmaurer! I will open a PR soon.

> Can you please do a PR and also add a comment in the code why it should never be removed again.

I will do this! 😄

**Comment by jasontedor** [comment]
I opened #6696.
